signalflow/client.go

```diff
@@ -0,0 +1,326 @@
+)
+
+// Client for SignalFlow via websockets (SSE is not currently supported).
+type Client struct {
+	// Access token for the org
+	token                  string
+	userAgent              string
+	defaultMetadataTimeout time.Duration
+	nextChannelNum         int64
+	conn                   *websocket.Conn
+	outgoingMessages       chan interface{}
+	readTimeout            time.Duration
+	// How long to wait for writes to the websocket to finish
+	writeTimeout   time.Duration
+	streamURL      *url.URL
+	channelsByName map[string]*Channel
+
+	isInitialized bool
+	isShutdown    bool
+	ctx           context.Context
+	cancel        context.CancelFunc
+	lastErr       error
+	lock          sync.Mutex
+}
+
+// ClientParam is the common type of configuration functions for the SignalFlow client
+type ClientParam func(*Client) error
+
+// StreamURL lets you set the full URL to the stream endpoint, including the
+// path.
+func StreamURL(streamEndpoint string) ClientParam {
+	return func(c *Client) error {
+		var err error
+		c.streamURL, err = url.Parse(streamEndpoint)
+		return err
+	}
+}
+
+// StreamURLForRealm can be used to configure the websocket url for a specific
+// SignalFx realm.
+func StreamURLForRealm(realm string) ClientParam {
+	return func(c *Client) error {
+		var err error
+		c.streamURL, err = url.Parse(fmt.Sprintf("wss://stream.%s.signalfx.com/v2/signalflow", realm))
+		return err
+	}
+}
+
+// AccessToken can be used to provide a SignalFx organization access token or
+// user access token to the SignalFlow client.
+func AccessToken(token string) ClientParam {
+	return func(c *Client) error {
+		c.token = token
+		return nil
+	}
+}
+
+// UserAgent allows setting the `userAgent` field when authenticating to
+// SignalFlow.  This can be useful for accounting how many jobs are started
+// from each client.
+func UserAgent(userAgent string) ClientParam {
+	return func(c *Client) error {
+		c.userAgent = userAgent
+		return nil
+	}
+}
+
+// MetadataTimeout is the default amount of time that calls to metadata
+// accessors on a SignalFlow Computation instance will wait to receive the
+// metadata from the backend before failing and returning a zero value. Usually
+// metadata comes in very quickly from the stream after the job start.
+func MetadataTimeout(timeout time.Duration) ClientParam {
+	return func(c *Client) error {
+		c.defaultMetadataTimeout = timeout
+		return nil
+	}
+}
+
+// ReadTimeout sets the duration to wait between messages that come on the
+// websocket.  If the resolution of the job is very low, this should be
+// increased.
+func ReadTimeout(timeout time.Duration) ClientParam {
+	return func(c *Client) error {
+		c.readTimeout = timeout
+		return nil
+	}
+}
+
+// NewClient makes a new, but uninitialized, SignalFlow client.
+func NewClient(options ...ClientParam) (*Client, error) {
+	c := &Client{
+		streamURL: &url.URL{
+			Scheme: "wss",
+			Host:   "stream.us0.signalfx.com",
```

Comment:
I assume we're ok to not add a convenience function for realm, since this handles that and more…

Reply:
The main client (not the signalflow one) should have a `Realm` option and then it gets passed through to the existing `StreamURLForRealm` option in the SignalFlow client.  That can wait for later though I suppose.